File: ta4j-core/src/main/java/org/ta4j/core/criteria/LinearTransactionCostCriterion.java
/*
 * SPDX-License-Identifier: MIT
 */
package org.ta4j.core.criteria;

import org.ta4j.core.BarSeries;
import org.ta4j.core.Position;
import org.ta4j.core.TradeView;
import org.ta4j.core.TradingRecord;
import org.ta4j.core.criteria.ReturnRepresentation;
import org.ta4j.core.criteria.pnl.GrossReturnCriterion;
import org.ta4j.core.num.Num;

/**
 * A linear transaction cost criterion.
 *
 * <p>
 * Calculates the transaction cost according to an initial traded amount and a
 * linear function defined by a and b (a * x + b).
 */
public class LinearTransactionCostCriterion extends AbstractAnalysisCriterion {

    private final double initialAmount;

    private final double a;
    private final double b;

    private final GrossReturnCriterion grossReturn;

    /**
     * Constructor. (a * x)
     *
     * @param initialAmount the initially traded amount
     * @param a             the a coefficient (e.g. 0.005 for 0.5% per
     *                      {@link TradeView trade})
     */
    public LinearTransactionCostCriterion(double initialAmount, double a) {
        this(initialAmount, a, 0);
    }

    /**
     * Constructor. (a * x + b)
     *
     * @param initialAmount the initially traded amount
     * @param a             the a coefficient (e.g. 0.005 for 0.5% per
     *                      {@link TradeView trade})
     * @param b             the b constant (e.g. 0.2 for $0.2 per {@link TradeView
     *                      trade})
     */
    public LinearTransactionCostCriterion(double initialAmount, double a, double b) {
        this.initialAmount = initialAmount;
        this.a = a;
        this.b = b;
        // Use MULTIPLICATIVE (1-based) for internal calculations since we multiply
        // amounts
        grossReturn = new GrossReturnCriterion(ReturnRepresentation.MULTIPLICATIVE);
    }

    @Override
    public Num calculate(BarSeries series, Position position) {
        return getTradeCost(series, position, series.numFactory().numOf(initialAmount));
    }

    @Override
    public Num calculate(BarSeries series, TradingRecord tradingRecord) {
        Num totalCosts = series.numFactory().zero();
        Num tradedAmount = series.numFactory().numOf(initialAmount);

        for (Position position : tradingRecord.getPositions()) {
            Num tradeCost = getTradeCost(series, position, tradedAmount);
            totalCosts = totalCosts.plus(tradeCost);
            // To calculate the new traded amount:
            // - Remove the cost of the *first* trade
            // - Multiply by the profit ratio
            // - Remove the cost of the *second* trade
            tradedAmount = tradedAmount.minus(getTradeCost(position.getEntry(), tradedAmount));
            tradedAmount = tradedAmount.multipliedBy(grossReturn.calculate(series, position));
            tradedAmount = tradedAmount.minus(getTradeCost(position.getExit(), tradedAmount));
        }

        // Special case: if the current position is open
        Position currentPosition = tradingRecord.getCurrentPosition();
        if (currentPosition.isOpened()) {
            totalCosts = totalCosts.plus(getTradeCost(currentPosition.getEntry(), tradedAmount));
        }

        return totalCosts;
    }

    /** The lower the criterion value, the better. */
    @Override
    public boolean betterThan(Num criterionValue1, Num criterionValue2) {
        return criterionValue1.isLessThan(criterionValue2);
    }

    /**
     * @param trade        the trade
     * @param tradedAmount the amount of the trade
     * @return the absolute trade cost
     */
    private Num getTradeCost(TradeView trade, Num tradedAmount) {
        final var numFactory = tradedAmount.getNumFactory();
        Num tradeCost = numFactory.zero();
        if (trade != null) {
            return numFactory.numOf(a).multipliedBy(tradedAmount).plus(numFactory.numOf(b));
        }
        return tradeCost;
    }

    /**
     * @param series        the bar series
     * @param position      the position
     * @param initialAmount the initially traded amount for the position
     * @return the absolute total cost of all trades in the position
     */
    private Num getTradeCost(BarSeries series, Position position, Num initialAmount) {
        Num totalTradeCost = series.numFactory().zero();
        if (position != null && position.getEntry() != null) {
            totalTradeCost = getTradeCost(position.getEntry(), initialAmount);
            if (position.getExit() != null) {
                // To calculate the new traded amount:
                // - Remove the cost of the first trade
                // - Multiply by the profit ratio
                Num newTradedAmount = initialAmount.minus(totalTradeCost)
                        .multipliedBy(grossReturn.calculate(series, position));
                totalTradeCost = totalTradeCost.plus(getTradeCost(position.getExit(), newTradedAmount));
            }
        }
        return totalTradeCost;
    }
}

File: ta4j-core/src/main/kotlin/org/ta4j/core/criteria/LinearTransactionCostCriterion.kt
/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2017-2022 Ta4j Organization & respective
 * authors (see AUTHORS)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package org.ta4j.core.criteria

import org.ta4j.core.*
import org.ta4j.core.criteria.pnl.GrossReturnCriterion
import org.ta4j.core.num.*

/**
 * A linear transaction cost criterion.
 *
 * Calculates the transaction cost according to an initial traded amount and a
 * linear function defined by a and b (a * x + b).
 */
/**
 * Constructor. (a * x + b)
 *
 * @param initialAmount the initially traded amount
 * @param a             the a coefficient (e.g. 0.005 for 0.5% per [                      trade][Trade])
 * @param b             the b constant (e.g. 0.2 for $0.2 per [                      trade][Trade])
 */


class LinearTransactionCostCriterion @JvmOverloads constructor(
    private val initialAmount: Double,
    private val a: Double,
    private val b: Double = 0.0
) : AbstractAnalysisCriterion() {
    private val grossReturn = GrossReturnCriterion()

    override fun calculate(series: BarSeries, position: Position): Num {
        return getTradeCost(series, position, series.numOf(initialAmount))
    }

    override fun calculate(series: BarSeries, tradingRecord: TradingRecord): Num {
        var totalCosts = series.numOf(0)
        var tradedAmount = series.numOf(initialAmount)
        for (position in tradingRecord.positions) {
            val tradeCost = getTradeCost(series, position, tradedAmount)
            totalCosts = totalCosts.plus(tradeCost)
            // To calculate the new traded amount:
            // - Remove the cost of the *first* trade
            // - Multiply by the profit ratio
            // - Remove the cost of the *second* trade
            tradedAmount = tradedAmount.minus(getTradeCost(position.entry, tradedAmount))
            tradedAmount = tradedAmount.times(grossReturn.calculate(series, position))
            tradedAmount = tradedAmount.minus(getTradeCost(position.exit, tradedAmount))
        }

        // Special case: if the current position is open
        val currentPosition = tradingRecord.getCurrentPosition()
        if (currentPosition.isOpened) {
            totalCosts = totalCosts.plus(getTradeCost(currentPosition.entry, tradedAmount))
        }
        return totalCosts
    }

    /** The lower the criterion value, the better.  */
    override fun betterThan(criterionValue1: Num, criterionValue2: Num): Boolean {
        return criterionValue1.isLessThan(criterionValue2)
    }

    /**
     * @param trade        the trade
     * @param tradedAmount the amount of the trade
     * @return the absolute trade cost
     */
    private fun getTradeCost(trade: Trade?, tradedAmount: Num?): Num {
        val tradeCost = tradedAmount!!.numOf(0)
        return if (trade != null) {
            tradedAmount.numOf(a).times(tradedAmount).plus(tradedAmount.numOf(b))
        } else tradeCost
    }

    /**
     * @param series        the bar series
     * @param position      the position
     * @param initialAmount the initially traded amount for the position
     * @return the absolute total cost of all trades in the position
     */
    private fun getTradeCost(series: BarSeries?, position: Position?, initialAmount: Num): Num {
        var totalTradeCost = series!!.numOf(0)
        if (position != null) {
            if (position.entry != null) {
                totalTradeCost = getTradeCost(position.entry, initialAmount)
                if (position.exit != null) {
                    // To calculate the new traded amount:
                    // - Remove the cost of the first trade
                    // - Multiply by the profit ratio
                    val newTradedAmount = initialAmount.minus(totalTradeCost)
                        .times(grossReturn.calculate(series, position))
                    totalTradeCost = totalTradeCost.plus(getTradeCost(position.exit, newTradedAmount))
                }
            }
        }
        return totalTradeCost
    }
}